### DISCUSSION TITLE
Ability to generate swagger/openapi file without running the app?

### DISCUSSION THREAD

**Discussion by mlasevich** [fastapi/fastapi#12216] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from contextlib import asynccontextmanager
import time

from fastapi import FastAPI


@asynccontextmanager
async def lifespan(app: FastAPI):
    """ Lifespan that mimics heavy/dependent startup/shutdown """
    print("starting...")
    time.sleep(60)
    print("started...")
    yield
    print("ending...")
    time.sleep(60)
    print("ended...")


app = FastAPI(lifespan=lifespan)


@app.get("/")
async def hello():
    """ hello world """
    return "Hello, World"
```

### Description

Maybe a silly question - but I would like to export the openapi.json file at build time, ideally without running the app. I get that I can run the app and go to the URL and download it, but starting up the app tends to require some background support processes and other scaffolding to support lifecycle events, etc. I am hoping to run something along the lines of `fastapi schema > openapi.json` (not necessarily using the cli, but the cli seems like a good tool for this)

Is this posible?

Simple usecases are to publish the swagger/openapi documentation on build, or  auto-generating a client library at build time, and so on.

### Operating System

Linux

### Operating System Details

Should not be relevant

### FastAPI Version

any

### Pydantic Version

n/a

### Python Version

3.11+

### Additional Context

N/A

**Comment by tristan** [comment, marked as answer]
You can call the `openapi()` method on the app without starting it and dump the results to a file.

e.g. you could have a separate module in your code that looks like this:

```python
import json
from your_app_module import app

with open("openapi.json", "w") as f:
    f.write(json.dumps(app.openapi()))
```

then running that you will get an `openapi.json` files (you can replace `json.dumps` with your encoder of choice, e.g. yaml).

**Comment by mlasevich** [comment]
I actually figured that out and submitted a [PR for fastapi-cli](https://github.com/fastapi/fastapi-cli/pull/91) to add a "schema" command. Looks like the PR failed some checks :-(
